packages/controls/src/controls/group/group.ts

```diff
@@ -0,0 +1,310 @@
+
+    const propsData = z.object(
+      mapValues(keys, (key) => key.data),
+    ) as SchemaType<PropsData<C>>
+
+    const versionedData = z.object({
+      [ControlDataTypeKey]: z.union([
+        z.literal(Definition.v1DataType),
+        z.literal(ShapeV2Definition.v1DataType),
+      ]),
+      value: propsData,
+    }) as SchemaType<VersionedData<C>>
+
+    const data = z.union([propsData, versionedData])
+
+    const value = z.object(mapValues(keys, (key) => key.value)) as SchemaType<
+      ValueType<C>
+    >
+
+    const resolvedValue = z.object(
+      mapValues(keys, (key) => key.resolvedValue),
+    ) as SchemaType<ResolvedValueType<C>>
+
+    const config = z.object({
+      label: z.string().optional(),
+      preferredLayout: z.union([
+        z.literal(Definition.Layout.Inline),
+        z.literal(Definition.Layout.Popover),
+      ]),
+      props: z.object(mapValues(keys, (key) => key.definition)),
+    })
+
+    const definition = z.object({
+      type,
+      config,
+    })
+
+    return { type, data, value, resolvedValue, definition, version }
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
+    return safeParse(this.schema.data, data)
+  }
+
+  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
+    if (data == null) return undefined
+
+    const propsData = Definition.propsData(data)
+
+    return mapValues(this.propDefs, (def, key) => def.fromData(propsData[key]))
+  }
+
+  propsToData(propsData: PropsData<C>): DataType<C> {
+    return {
+      ...this.dataSignature,
+      value: propsData,
+    }
+  }
+
+  toData(value: ValueType<C>): DataType<C> {
+    return this.propsToData(
+      mapValues(this.propDefs, (def, key) => def.toData(value[key])),
+    )
+  }
+
+  copyData(
+    data: DataType<C> | undefined,
+    context: CopyContext,
+  ): DataType<C> | undefined {
+    if (data == null) return undefined
+
+    const propsData = Definition.propsData(data)
+
+    return this.propsToData(
+      mapValues(this.propDefs, (def, key) =>
+        def.copyData(propsData[key], context),
+      ),
+    )
+  }
+
+  getTranslatableData(data: DataType<C>): Data {
+    if (data == null) return null
+
+    const propsData = Definition.propsData(data)
+
+    return mapValues(this.propDefs, (def, key) =>
+      def.getTranslatableData(propsData[key]),
+    )
+  }
+
+  mergeTranslatedData(
+    data: DataType<C>,
+    translatedData: Record<string, DataType<C>>,
+    context: MergeTranslatableDataContext,
+  ): Data {
+    if (translatedData == null) return data
+
+    const propsData = Definition.propsData(data)
+
+    return mapValues(this.propDefs, (def, key) =>
+      def.mergeTranslatedData(propsData[key], translatedData[key], context),
+    )
+  }
+
+  createInstance(sendMessage: SendMessage): InstanceType<C> {
+    return new GroupControl(this, sendMessage)
+  }
+
+  serialize(): [SerializedRecord, Transferable[]] {
+    return serialize(this.config, {
```

Comment:
Note that `this.dataType` is not serialized. Deserialization is based on the `Definition.type`, and deserialization for `ShapeV2` is handled [there](https://github.com/makeswift/makeswift/pull/909/files?diff=split&w=1#diff-0a3eea440467c5a2bac813267bc3b89cccda39463c408be806163f5b69c205cfR13).